### ISSUE TITLE
Qwen2.5VL is broken!

### ISSUE THREAD

**Issue by pengzhenghao** [huggingface/transformers#40154]
### System Info

- `transformers` version: 4.55.0
- Platform: Linux-5.15.0-1032-oracle-x86_64-with-glibc2.35
- Python version: 3.12.3
- Huggingface_hub version: 0.34.3
- Safetensors version: 0.6.1
- Accelerate version: 1.10.0
- Accelerate config:    not found
- DeepSpeed version: 0.17.4
- PyTorch version (accelerator?): 2.7.1+cu128 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA A100-SXM4-80GB

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

In recent PR: https://github.com/huggingface/transformers/pull/39447

A text position embedding is introduced:

https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1624-R1628

```python
        if position_ids.ndim == 3 and position_ids.shape[0] == 4:
            text_position_ids = position_ids[0]
            position_ids = position_ids[1:]
        else:
            text_position_ids = position_ids[0]
```

The `text_position_ids` will be fed to the decoder layer: https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1665

```python
            layer_outputs = decoder_layer(
                hidden_states,
                attention_mask=causal_mask_mapping[decoder_layer.attention_type],
                position_ids=text_position_ids,
                past_key_value=past_key_values,
                output_attentions=output_attentions,
                use_cache=use_cache,
```

OK. So far we have learned:

1. If `position_ids` inputs as a 3-dim tensor with shape (4, B, seq_len), then `text_position_ids` is the first (1, B, seq_len).
2. Otherwise, it just copy the first `position_ids[0]`

Now, you can easily verify that this value in training time is a 3-dim tensor with shape (3, B, seq_len), as shown in this figure: (note that I have visual input so seq_len != max(pos_id)

<img width="1278" height="436" alt="Image" src="https://github.com/user-attachments/assets/b87cf81b-ed3e-4b06-ac7d-55eb2f4586c4" />

Therefore, text_position_ids is simply (0, ..., 631) as you would expect in previous version.

However, in generation time, the text_position_ids is prepared by:

https://github.com/huggingface/transformers/pull/39447/files#diff-41bc5f79048d22eb30688e5440a89f4303ed02aa3d7c99aa807e3e7243e1fc0aR861-R864

```python
            if "position_ids" not in model_inputs:
                text_positions = torch.arange(input_ids, device=input_ids.device)[None, None, :]
            else:
                text_positions = model_inputs["position_ids"][None, ...]
```

If we take a look ahead, model_inputs["position_ids"] is prepared by GenerationMixin. In both case, text_positions is default to the (0, ..., seq_len-1). Then, this text_positions (which is already WRONG because it doesn't take care of the offset caused by potential visual tokens) will be concat to form a 4D position_ids.

```python
model_inputs["position_ids"] = torch.cat([text_positions, vision_positions], dim=0)
```

Then, such a value will be sent to decoder layer.

In summary, in training time, the decoder layer will see position_ids to be offsetted by visual token. But in generation, it see a plain (0,...,seq_len-1) position_ids.

This might be the root cause to https://github.com/huggingface/transformers/issues/40136

### Expected behavior

I expect it works normally.

**Comment by pengzhenghao** [comment]
BTW, I also found the model has some problem to generate anything with batch size > 1 (aka, bs>1 or you are using num_return_sequences>1) if using flash-attention 2. However, when using SDPA it would be fine. 

I have no idea what's going on, but according to https://github.com/huggingface/transformers/issues/40136#issuecomment-3185966650 seems like there are some issue in FA2 argument passing.

**Comment by zucchini-nlp** [comment]
> In summary, in training time, the decoder layer will see position_ids to be offsetted by visual token. But in generation, it see a plain (0,...,seq_len-1) position_ids.

Ah right, in train time the positions will not be correct for FA2 packing, and we can't expected all users to pass 4D positions. That must be the issue if you are using FA2 

And I will check the batched generation, no idea what might have gone wrong. EDIT: with simple images the generated text makes sense with bs=2. Do you have an example of broken inference?

**Comment by pengzhenghao** [comment]
@zucchini-nlp Thanks for getting back! I am investigating this and will get back soon.


One thing I've observed is `cu_seq_lens_q/k`:

<img width="614" height="41" alt="Image" src="https://github.com/user-attachments/assets/6c64f993-18cc-4a58-bb4c-f6bfad91eb4f" />


<img width="535" height="221" alt="Image" src="https://github.com/user-attachments/assets/8a39cc4c-7754-4871-8d7b-b3c49e996307" />

Seems like the cu_seq_lens_k is added by one and cause mismatch between cu_seq_lens_q.

<img width="560" height="275" alt="Image" src="https://github.com/user-attachments/assets/19538039-9501-4622-88e1-9dbc617ae7e0" />



Do you think this is a correct behavior?


Here I have bs=12 and seq_len is 2294. So I should expect to see both `cu_seq_lens_k` and `cu_seq_lens_q` to be:

`[0, 2294, 2294*2=4588, 2294*3=6882, ...]`

Is that correct? Then the cu_seq_lens_k is wrong.

**Comment by pengzhenghao** [comment]
<img width="671" height="294" alt="Image" src="https://github.com/user-attachments/assets/d20e5fa4-874f-4192-aefe-4b290b75143b" />

OK with the above changes my result is normal now. Lemme draft a PR.

So actually we've identified two issues:

1. The wrong text_input_ids to Qwen25VL (not sure the outreach of this, it happens to Qwen2VL? IDK)
2. The flash attention 2 varlen packing.

**Comment by pengzhenghao** [comment]
OK I found that issue 2 is fixed just 5 hours ago:

https://github.com/huggingface/transformers/pull/40161

**Comment by pengzhenghao** [comment]
For the issue 1, I just make a PR #40177
